### PR TITLE
adding container development env & ShubProvisioner

### PR DESCRIPTION
**Pull request by vsoch**
**Description of the Pull Request (PR):**

Hey @bauerm97 ! I stepped back and decided that before trying to implement the Shub endpoint feature, it would be reasonable to have a clean method to reproduce my development workflow, period. The easiest thing (I think) to do is to create a Dockerfile (to live with a respective development repository) where a developer can just build the image and go! (no pun intended, har har.)

** update ** have fixed previous issues!

**Building**

```bash
$ docker build -t vanessa/singularity-dev .
```

**Interactive Session**
```bash
$ docker run --privileged -it --entrypoint bash vanessa/singularity-dev
```

**Testing**
```bash
$ docker run --privileged -it --entrypoint go vanessa/singularity-dev test ./tests/
```

A few notes:
 - A more detailed description of these commands is in the header of the Dockerfile
 - I didn't see where the CHANGELOG went? I would want to add the note about the Dockerfile here.
 - The alpine base is the most common for go development, and I originally chose it. The make process failed, likely because of library mismatches, so I went back to debian. Maybe in the future we could revisit alphine because it's much smaller.
- if/when we get this working, there should be a place to write instructions for using it. The logical place would be Singularity docs, and I can start with a markdown post somewhere,
- Derivations of the above can do things like bind the local directory to /code, etc.

Now I'll work on that other thing :)

### CODE REVIEW

**Comment by vsoch** [comment]
okay just updated the notes! I can definitely add a Singularity recipe too! Then we will have the ultimate matrix, and lazy developers like me will be very happy :)

File: src/pkg/build/provisioners/provisioner_shub.go [hunk 1]
```diff
@@ -9,23 +9,143 @@
...
+	err = json.NewDecoder(res.Body).Decode(&p)
+	return
+
+}
+
+// NewShubProvisioner returns a provisioner that can dump a previously
+// build Singularity image into a sandbox based on a URI
+func NewShubProvisioner(src string) (p *ShubProvisioner, err error) {
+
+	// Question for bauerm - where does this get called?
```
Comment (bauerm97) [inline src/pkg/build/provisioners/provisioner_shub.go]: `NewShubProvisioner` gets called in `src/pkg/build/provisioner.go`, if that was the question :-)

File: src/pkg/build/provisioners/provisioner_shub.go [hunk 2]
```diff
@@ -9,23 +9,143 @@
...
+	oci "github.com/containers/image/oci/layout"
 	"github.com/singularityware/singularity/src/pkg/image"
+	"github.com/singularityware/singularity/src/pkg/sylog"
 )
 
-// ShubProvisioner provisions a sandbox environment from a shub
-// URI.
+
+// ShubClient contains the uri and client
+type ShubClient struct {
```
Comment (bauerm97) [inline src/pkg/build/provisioners/provisioner_shub.go]: This probably should not be an exported type

File: src/pkg/build/provisioners/provisioner_shub.go [hunk 3]
```diff
@@ -9,23 +9,143 @@
...
+	Client   http.Client
+	ImageUri string
+	HTTPAddr string
+}
+
+// NewRemoteBuilder creates a RemoteBuilder with the specified details.
+func NewShubClient(uri string) (shubClient *ShubClient) {
+
+	sylog.Infof("%v\n", uri)
+	httpAddr := fmt.Sprintf("www.singularity-hub.org")
```
Comment (bauerm97) [inline src/pkg/build/provisioners/provisioner_shub.go]: `fmt.Sprintf` probably unnecessary here. You can just change: 

`HTTPAddr: httpAddr,` 

on line 47 to:

`HTTPAddr: "www.singularity-hub.org",`

Go is, in general, much more lenient w.r.t. strings and memory allocation than C

File: src/pkg/build/provisioners/provisioner_shub.go [hunk 4]
```diff
@@ -9,23 +9,143 @@
...
+
+        // TODO: need to parse the tag / digest and send along too
+        uri := strings.Split(p.srcRef.StringWithinTransport(), ":")[0]
+
+	// Format the http address, coinciding with the image uri
+	httpAddr := fmt.Sprintf("www.singularity-hub.org/api/container%s/", uri)
+	sylog.Infof("%v\n", httpAddr)
+
+	// Create the request, add headers context
+	url := url.URL{Scheme: "https", Host: sc.HTTPAddr, Path: httpAddr}
```
Comment (bauerm97) [inline src/pkg/build/provisioners/provisioner_shub.go]: In general when instantiating struct objects it's cleaner to do each field on it's own line (for readability and consistency sake):

```
url := url.URL{
    Scheme: "https",
    Host: sc.HTTPAddr,
    Path: httpAddr,
}
```

**Review by bauerm97 — COMMENTED** [review]
Just a few general comments after a quick read through. Looking good so far V!

File: src/pkg/build/provisioners/provisioner_shub.go (hunk 1 above)
Comment (vsoch) [reply src/pkg/build/provisioners/provisioner_shub.go]: yes, thank you!

File: src/pkg/build/provisioners/provisioner_shub.go (hunk 2 above)
Comment (vsoch) [reply src/pkg/build/provisioners/provisioner_shub.go]: I didn't know this was a thing, haha. Reading now - holy cow you just capitalize and it becomes exported? Nuts! Agreed here, I will uncapitalize.

File: src/pkg/build/provisioners/provisioner_shub.go (hunk 3 above)
Comment (vsoch) [reply src/pkg/build/provisioners/provisioner_shub.go]: okay, I've adjusted this one step further, I put it directly into the data structure:

```
	sc = &shubClient{
		Client: http.Client{
			Timeout: 30 * time.Second,
		},
		HTTPAddr: "www.singularity-hub.org",
		ImageUri: uri,
	}

```

File: src/pkg/build/provisioners/provisioner_shub.go (hunk 4 above)
Comment (vsoch) [reply src/pkg/build/provisioners/provisioner_shub.go]: Agree.

**Comment by vsoch** [comment]
okay here is an update on where we are - I'm working on parsing the manifest.

```bash
r# singularity build /tmp/test.simg shub://vsoch/hello-world
INFO    [U=0,P=11841]      getManifest()                 www.singularity-hub.org/api/container//vsoch/hello-world/
INFO    [U=0,P=11841]      getManifest()                 {https    www.singularity-hub.org/api/container//vsoch/hello-world/  false  }
INFO    [U=0,P=11841]      getManifest()                 response: &{200 OK 200 HTTP/1.1 1 1 map[Connection:[keep-alive] Vary:[Accept, Cookie] Allow:[GET, HEAD, OPTIONS] X-Frame-Options:[SAMEORIGIN] Server:[nginx/1.13.5] Date:[Mon, 07 May 2018 17:30:35 GMT] Content-Type:[application/json] Content-Length:[464]] 0xc4202c8100 464 [] false false map[] 0xc420134000 0xc420021600}
INFO    [U=0,P=11841]      getManifest()                 body: [123 34 105 100 34 58 50 51 44 34 110 97 109 101 34 58 34 118 115 111 99 104 47 104 101 108 108 111 45 119 111 114 108 100 34 44 34 105 109 97 103 101 34 58 34 104 116 116 112 115 58 47 47 119 119 119 46 103 111 111 103 108 101 97 112 105 115 46 99 111 109 47 100 111 119 110 108 111 97 100 47 115 116 111 114 97 103 101 47 118 49 47 98 47 115 105 110 103 117 108 97 114 105 116 121 104 117 98 47 111 47 115 105 110 103 117 108 97 114 105 116 121 104 117 98 37 50 70 103 105 116 104 117 98 46 99 111 109 37 50 70 118 115 111 99 104 37 50 70 104 101 108 108 111 45 119 111 114 108 100 37 50 70 101 50 55 57 52 51 50 101 54 100 51 57 54 50 55 55 55 98 98 55 98 53 101 56 100 53 52 102 51 48 102 52 51 52 55 100 56 54 55 101 37 50 70 101 100 57 55 53 53 97 48 56 55 49 102 48 52 100 98 51 101 49 52 57 55 49 98 101 99 53 54 97 51 51 102 37 50 70 101 100 57 55 53 53 97 48 56 55 49 102 48 52 100 98 51 101 49 52 57 55 49 98 101 99 53 54 97 51 51 102 46 115 105 109 103 63 103 101 110 101 114 97 116 105 111 110 61 49 53 48 56 48 55 50 48 50 53 53 56 57 56 50 48 38 97 108 116 61 109 101 100 105 97 34 44 34 98 114 97 110 99 104 34 58 34 109 97 115 116 101 114 34 44 34 99 111 109 109 105 116 34 58 34 101 50 55 57 52 51 50 101 54 100 51 57 54 50 55 55 55 98 98 55 98 53 101 56 100 53 52 102 51 48 102 52 51 52 55 100 56 54 55 101 34 44 34 116 97 103 34 58 34 108 97 116 101 115 116 34 44 34 118 101 114 115 105 111 110 34 58 34 101 100 57 55 53 53 97 48 56 55 49 102 48 52 100 98 51 101 49 52 57 55 49 98 101 99 53 54 97 51 51 102 34 44 34 115 105 122 101 95 109 98 34 58 34 51 51 51 34 125]
INFO    [U=0,P=11841]      getManifest()                 manifest: &{   }
...
above is early attempts to parse the manifest, not sure I'm getting it yet!
below the break is where I'll write the part to download the image itself, and move (extract?) to sandbox.
...
INFO    [U=0,P=11841]      Provision()                   <nil>Parallel mksquashfs: Using 4 processors
Creating 4.0 filesystem on /tmp/squashfs-520993920.img, block size 131072.


Exportable Squashfs 4.0 filesystem, gzip compressed, data block size 131072
	compressed data, compressed metadata, compressed fragments, compressed xattrs
	duplicates are removed
Filesystem size 0.15 Kbytes (0.00 Mbytes)
	99.37% of uncompressed filesystem size (0.15 Kbytes)
Inode table size 29 bytes (0.03 Kbytes)
	85.29% of uncompressed inode table size (34 bytes)
Directory table size 0 bytes (0.00 Kbytes)
	-nan% of uncompressed directory table size (0 bytes)
Number of duplicate files found 0
Number of inodes 1
Number of files 0
Number of fragments 0
Number of symbolic links  0
Number of device nodes 0
Number of fifo nodes 0
Number of socket nodes 0
Number of directories 1
Number of ids (unique uids + gids) 1
Number of uids 1
	root (0)
Number of gids 1
	root (0)

Build Succeeded
```

**Comment by vsoch** [comment]
Some random things I've inferred that I want to check:

when a function just has return, it isn't returning nil necessarily, but the variable specified in its declaration. For example:

```
func NewShubProvisioner(src string) (p *ShubProvisioner, err error) {
```
if I just do:
```
return
```
it would return p, err ? But I could also do it explicitly
```
return p, nil
```
So TLDR: GoLang requires you being very specific about variables not just as inputs, but as outputs too. I would imagine it's hard to have a function that returns something general (or different things in different cases)

-------------------------------------------------------------------------------------------------------------------------------------

When you have a function called NewX it is the function that defines how to create a new object of type X. If you don't have a NewX function then do you need to use new explicitly, like:

```bash
var manifest = new(shubApiResponse)
```
----------------------------------------------------------------------------------------------------------------------------
oh woot! Just found my manifest image url :) Will do some more futzing around and push more changes after that!
```
INFO    [U=0,P=13456]      getManifest()                 manifest: &{https://www.googleapis.com/download/storage/v1/b/singularityhub/o/singularityhub%2Fgithub.com%2Fvsoch%2Fhello-world%2Fe279432e6d3962777bb7b5e8d54f30f4347d867e%2Fed9755a0871f04db3e14971bec56a33f%2Fed9755a0871f04db3e14971bec56a33f.simg?generation=1508072025589820&alt=media vsoch/hello-world latest ed9755a0871f04db3e14971bec56a33f}
```

**Comment by vsoch** [comment]
okay download of file - I'm just dumping it into the same temp-oci directory (to use the same convention as Docker)
```bash
root@0c5a31c18cea:/go/src/github.com/singularityware/singularity/builddir# singularity build /tmp/test.simg shub://vsoch/hello-world
INFO    [U=0,P=25181]      getManifest()                 www.singularity-hub.org/api/container//vsoch/hello-world/
INFO    [U=0,P=25181]      getManifest()                 {https    www.singularity-hub.org/api/container//vsoch/hello-world/  false  }
INFO    [U=0,P=25181]      getManifest()                 response: &{200 OK 200 HTTP/1.1 1 1 map[Connection:[keep-alive] Vary:[Accept, Cookie] Allow:[GET, HEAD, OPTIONS] X-Frame-Options:[SAMEORIGIN] Server:[nginx/1.13.5] Date:[Mon, 07 May 2018 18:05:32 GMT] Content-Type:[application/json] Content-Length:[464]] 0xc4202cbc00 464 [] false false map[] 0xc420134100 0xc42053ad10}
INFO    [U=0,P=25181]      getManifest()                 manifest: https://www.googleapis.com/download/storage/v1/b/singularityhub/o/singularityhub%2Fgithub.com%2Fvsoch%2Fhello-world%2Fe279432e6d3962777bb7b5e8d54f30f4347d867e%2Fed9755a0871f04db3e14971bec56a33f%2Fed9755a0871f04db3e14971bec56a33f.simg?generation=1508072025589820&alt=media
INFO    [U=0,P=25181]      Provision()                   https://www.googleapis.com/download/storage/v1/b/singularityhub/o/singularityhub%2Fgithub.com%2Fvsoch%2Fhello-world%2Fe279432e6d3962777bb7b5e8d54f30f4347d867e%2Fed9755a0871f04db3e14971bec56a33f%2Fed9755a0871f04db3e14971bec56a33f.simg?generation=1508072025589820&alt=mediaINFO    [U=0,P=25181]      fetch()                       
Created temporary file /tmp/temp-oci-093757454/shub-image916890645
Parallel mksquashfs: Using 4 processors
Creating 4.0 filesystem on /tmp/squashfs-685410928.img, block size 131072.
```

**Comment by vsoch** [comment]
okay I am ready for some help! Right now the image downloads, and it's sitting in a temporary directory. In the Docker flow we would call a function (from their respective package) to extract the layers. I'm guessing we need to do the same for the singularity sif (or other, if older) image, and that maybe there are internal functions that already exist to do it? @bauerm97 here is the print output:

```bash
root@0c5a31c18cea:/go/src/github.com/singularityware/singularity/builddir# singularity build /tmp/test.simg shub://vsoch/hello-world
INFO    [U=0,P=2242]       getManifest()                 www.singularity-hub.org/api/container//vsoch/hello-world/
INFO    [U=0,P=2242]       getManifest()                 {https    www.singularity-hub.org/api/container//vsoch/hello-world/  false  }
INFO    [U=0,P=2242]       getManifest()                 response: &{200 OK 200 HTTP/1.1 1 1 map[Server:[nginx/1.13.5] Date:[Mon, 07 May 2018 18:21:51 GMT] Content-Type:[application/json] Content-Length:[464] Connection:[keep-alive] Vary:[Accept, Cookie] Allow:[GET, HEAD, OPTIONS] X-Frame-Options:[SAMEORIGIN]] 0xc4204fa180 464 [] false false map[] 0xc420128000 0xc42053cdc0}
INFO    [U=0,P=2242]       getManifest()                 manifest: https://www.googleapis.com/download/storage/v1/b/singularityhub/o/singularityhub%2Fgithub.com%2Fvsoch%2Fhello-world%2Fe279432e6d3962777bb7b5e8d54f30f4347d867e%2Fed9755a0871f04db3e14971bec56a33f%2Fed9755a0871f04db3e14971bec56a33f.simg?generation=1508072025589820&alt=media
INFO    [U=0,P=2242]       Provision()                   https://www.googleapis.com/download/storage/v1/b/singularityhub/o/singularityhub%2Fgithub.com%2Fvsoch%2Fhello-world%2Fe279432e6d3962777bb7b5e8d54f30f4347d867e%2Fed9755a0871f04db3e14971bec56a33f%2Fed9755a0871f04db3e14971bec56a33f.simg?generation=1508072025589820&alt=media
INFO    [U=0,P=2242]       fetch()                       
Creating temporary image file /tmp/temp-shub-752751900/shub-container935023563
INFO    [U=0,P=2242]       unpackTmpfs()                 Here we need to unpack /tmp/temp-shub-752751900/shub-container935023563 to &{/tmp/test.simg-952661105}
Parallel mksquashfs: Using 4 processors
Creating 4.0 filesystem on /tmp/squashfs-724622510.img, block size 131072.

Exportable Squashfs 4.0 filesystem, gzip compressed, data block size 131072
	compressed data, compressed metadata, compressed fragments, compressed xattrs
	duplicates are removed
Filesystem size 0.15 Kbytes (0.00 Mbytes)
	99.37% of uncompressed filesystem size (0.15 Kbytes)
Inode table size 29 bytes (0.03 Kbytes)
	85.29% of uncompressed inode table size (34 bytes)
Directory table size 0 bytes (0.00 Kbytes)
	-nan% of uncompressed directory table size (0 bytes)
Number of duplicate files found 0
Number of inodes 1
Number of files 0
Number of fragments 0
Number of symbolic links  0
Number of device nodes 0
Number of fifo nodes 0
Number of socket nodes 0
Number of directories 1
Number of ids (unique uids + gids) 1
Number of uids 1
	root (0)
Number of gids 1
	root (0)


Build Succeeded
```

And of course see where it says:

```bash
INFO    [U=0,P=2242]       unpackTmpfs()                 Here we need to unpack /tmp/temp-shub-
```
that is where we need to...do that! Do you know if there is an already existing function in singularity to run (internally) the build from another image? I prepared the function to use it:

```bash
func (p *ShubProvisioner) unpackTmpfs(tmpfile string, i *image.Sandbox) (err error) {
	//refs := []string{"name=tmp"}
        sylog.Infof("Here we need to unpack %v to %v\n", tmpfile, i)
	//err = imagetools.UnpackLayout(p.tmpfs, i.Rootfs(), "amd64", refs)
	return err
}
```

**Pull request closed without merging**